### PR TITLE
Fix cube fitting compatibility with unit conversion

### PR DESCRIPTION
**Pull request by rosteen**
Opening as draft since I still need to work on the unit conversion part, but I wanted to see the diff and remind myself what I'd changed so far.

### CODE REVIEW

**Comment by cshanahan1** [comment]
Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:

`cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)`

**Review by cshanahan1 — COMMENTED** [review]
Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:

cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)

**Comment by rosteen** [comment]
> Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:
> 
> cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)

Hm, the model fit plots fine with the example notebook data, but now that you mention it I don't see table results. I'll look into it on Monday.

**Comment by rosteen** [comment]
> Tested this out, I do see the fix that now the component units are always in 'surface brightness' if the cube is surface brightness, and do not change based on what is selected for the spectral y axis, so that's fixed. However, I'm still not getting any results in the table or plotted when I run a cube fit on this loaded data using one of the test fixtures to create a cube:
> 
> cube = _create_spectrum1d_cube_with_fluxunit(shape=(25, 25, 4), fluxunit=u.Jy / u.sr, with_uncerts=True)

From a comment in the code: `# cube fits are currently unsupported in tables` - I think that's out of scope for this. I'll investigate the test fixture not plotting.

**Comment by rosteen** [comment]
https://github.com/user-attachments/assets/9bfe567d-39ae-47f4-abbd-a1bc35b3f9a4

@cshanahan1  hm, I see the model plot fine with that cube (I fit a constant to it). Are you sure you didn't need to go into the data menu and add the model to the viewer? Or did it fail to appear with the `Plot in Viewer` option toggled on?

 I do see an error trying to add a Linear model component, trying to fix that...

**Comment by rosteen** [comment]
Note that this merging this PR on top of #3156 resolves the test failures (other than the known axis label failure there).

**Comment by cshanahan1** [comment]
Screenshot 2024-09-16 at 1 49 02 PM
still seeing some out of sync units. I loaded a Jy/sr cube, changed the flux unit to MJy, and the model fitting plugin shows Jy as the flux unit. This happened when MJy was set before creating the component for the first time, so I shouldn't have to re-estimate parameters, but even when doing that the incorrect unit remains

**Comment by camipacifici** [comment]
Linear fitting to cube or single spectrum does not break anymore!
In this PR, I see the icon to open the model fitting table in a different window, but the link goes to a 404 window.

**Comment by rosteen** [comment]
> still seeing some out of sync units. I loaded a Jy/sr cube, changed the flux unit to MJy, and the model fitting plugin shows Jy as the flux unit. This happened when MJy was set before creating the component for the first time, so I shouldn't have to re-estimate parameters, but even when doing that the incorrect unit remains


Fixed, toggling the cube fit was triggering the dataset selected logic, which was resetting the units. I removed all the unit setting from there in favor of setting it when the first component is initialized (and whenever user changes units or toggles cube fit).

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py [hunk 1]
```diff
@@ -470,6 +465,17 @@ def _initialize_model_component(self, model_comp, comp_label, poly_order=None):
...
+        # Need to set the units the first time we initialize a model component, after this
+        # we listen for display unit changes
+        if (self._units is None or self._units == {} or 'x' not in self._units or
+                'y' not in self._units):
+            uc = self.app._jdaviz_helper.plugins['Unit Conversion']
+            self._units['x'] = uc.spectral_unit.selected
+            if self.cube_fit:
+                self._units['y'] = uc._obj.sb_unit_selected
+            else:
+                self._units['y'] = uc.flux_unit.selected
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:477]: eventually this probably will be safe, but might need to use `app._get_display_units` for now to get the fallbacks when unit conversion does not exist

**Comment by cshanahan1** [comment]
I see it now respecting flux unit selection when i change between Jy/MJy, but breaks on erg / (Angstrom s cm2 sr)

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py (hunk 1 above)
Comment (kecnry) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:477]: for non-cube fit, this also needs to listen to `spectral_y_type` (`app._get_display_units('spectral_y')` handles that logic as well).

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py [hunk 2]
```diff
@@ -309,6 +299,30 @@ def _warn_if_no_equation(self):
...
+            self._units['y'] = self.app._get_display_unit('sb')
+            self.dataset.add_filter('is_flux_cube')
+            self.dataset.remove_filter('layer_in_spectrum_viewer')
+        else:
+            self._units['y'] = self.app._get_display_unit('spectral_y')
+            self.dataset.add_filter('layer_in_spectrum_viewer')
+            self.dataset.remove_filter('is_flux_cube')
+
+        self.dataset._clear_cache()
+        self.reestimate_model_parameters()
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:324]: If we remove `reestimate_model_parameters` from here does this still work (just sometimes requiring the user to hit the button)?

I wonder if its ever useful to keep the estimates from the non cube fit as the starting point (especially if there is no unit conflict)?  Or if the user manually entered there own values but then they were changed by toggling cube fit, would that be unexpected/disruptive?

Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py]: Updated to only reestimate if `spectral_y` and `sb` have different units.

**Comment by rosteen** [comment]
Based on offline feedback from @camipacifici I changed the cube fit toggle to display the incompatible units warning, prompting the user to click to reestimate the parameters rather than doing it automatically. A future improvement will be to automatically convert using the scale factor if possible.

**Comment by rosteen** [comment]
Oh right - the test coverage is going to look bad because there are some getting skipped that need #3156.

**Comment by cshanahan1** [comment]
> display the incompatible units warning, prompting the user to click to reestimate the parameters rather than doing it automatically

I don't see this working. Using cubeviz example notebook, If I start in MJy and add a constant component and fit and plot a cube fit, then change the unit, the component is still in MJy with no warning. Interestingly, the fit plotted in the viewer after the unit change has the correct units, but the model component displayed does not

**Comment by cshanahan1** [comment]
Could you add a test that tests cube fitting with unit conversion(s)? Maybe something where you change from Jy > MJy, and Jy > er/s/cm2/A ?

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py [hunk 3]
```diff
@@ -309,6 +299,36 @@ def _warn_if_no_equation(self):
...
+        else:
+            # only want spectral viewers in the options
+            self.add_results.viewer.filters = ['is_spectrum_viewer']
+
+    @observe('cube_fit')
+    def _cube_fit_changed(self, event={}):
+        self._update_viewer_filters(event=event)
+
+        sb_unit = self.app._get_display_unit('sb')
+        spectral_y_unit = self.app._get_display_unit('spectral_y')
```
[the remark this answers is not in the document]
Comment (rosteen) [reply jdaviz/configs/default/plugins/model_fitting/model_fitting.py:315]: I'm using it a few lines down to check to see if the collapsed spectrum is in flux or surface brightness, to see if the parameter units will be compatible when toggling `cube_fit`.

**Comment by rosteen** [comment]
@cshanahan1 I added a test, but I won't be able to test changing the unit to `erg / (Angstrom s cm2)` until the contour bug is fixed.

**Comment by cshanahan1** [comment]
Tested this one final time and couldn't break it.  Talked on slack with ricky about a few points of confusion i had and he cleared those up so I'll approve, looks good!

**Review by cshanahan1 — APPROVED** [review]

File: jdaviz/configs/default/plugins/model_fitting/model_fitting.py [hunk 4]
```diff
@@ -930,7 +984,8 @@ def _fit_model_to_cube(self, add_data):
                 models_to_fit,
                 self.model_equation,
                 run_fitter=True,
-                window=None
+                window=None,
+                n_cpu=1  # Remove this after debugging!
```
Comment (cshanahan1) [inline jdaviz/configs/default/plugins/model_fitting/model_fitting.py:988]: pssst

**Review by gibsongreen — APPROVED** [review]
Did one last look over and round of testing, all looks good to me and works as I'd expect!

**Pull request merged**